iceoryx_dds/source/iceoryx_dds/dds/fastdds_data_reader.cpp

```diff
@@ -0,0 +1,287 @@
+        m_dataReader = m_subscriber->create_datareader(m_topic, dataReaderQos, &m_listener);
+
+        // Save state
+        m_isConnected.store(true, std::memory_order_relaxed);
+        LogDebug() << m_logName << "Connected to topic: " << topicString;
+    }
+}
+
+iox::cxx::optional<uint32_t> iox::dds::FastDDSDataReader::peekNextSize()
+{
+    // Get information about the next unread sample
+    Mempool::Chunk data{0, nullptr};
+    eprosima::fastdds::dds::SampleInfo nextInfo;
+    if (eprosima::fastrtps::types::ReturnCode_t::RETCODE_OK ==
+        m_dataReader->read_next_sample(&data ,&nextInfo))
+    {
+        if (data.size != 0)
+        {
+            return iox::cxx::optional<uint32_t>(static_cast<uint64_t>(data.size));
+        }
+    }
+
+    // no valid samples available
+    return iox::cxx::nullopt_t();
+}
+
+bool iox::dds::FastDDSDataReader::hasNewSamples()
+{
+    eprosima::fastdds::dds::SampleInfo nextInfo;
+    return (m_dataReader->get_first_untaken_info(&nextInfo) == eprosima::fastrtps::types::ReturnCode_t::RETCODE_OK);
+}
+
+iox::cxx::expected<iox::dds::DataReaderError> iox::dds::FastDDSDataReader::takeNext(
+        uint8_t* const buffer,
+        const uint64_t& bufferSize)
+{
+    // validation checks
+    if (!m_isConnected.load())
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::NOT_CONNECTED);
+    }
+    if (buffer == nullptr)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_RECV_BUFFER);
+    }
+
+    // take next sample and copy into buffer
+    Mempool::Chunk data{static_cast<uint32_t>(bufferSize), buffer};
+    eprosima::fastdds::dds::SampleInfo nextInfo;
+    eprosima::fastrtps::types::ReturnCode_t ret = m_dataReader->take_next_sample(&data, &nextInfo);
+    if (ret == eprosima::fastrtps::types::ReturnCode_t::RETCODE_NO_DATA)
+    {
+        // no samples available
+        return iox::cxx::success<>();
+    }
+    else if (ret != eprosima::fastrtps::types::ReturnCode_t::RETCODE_OK)
+    {
+        LogError() << m_logName << "take_next_sample returned and error";
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_DATA);
+    }
+
+    // valid size
+    auto sampleSize = data.size;
+    if (sampleSize == 0)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_DATA);
+    }
+    if (bufferSize < sampleSize)
+    {
+        // provided buffer is too small.
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::RECV_BUFFER_TOO_SMALL);
+    }
+
+    // copy data into the provided buffer
+    auto bytes = data.payload;
+    std::copy(bytes, bytes + sampleSize, const_cast<uint8_t*>(buffer));
+
+    return iox::cxx::success<>();
+}
+
+iox::cxx::expected<uint64_t, iox::dds::DataReaderError> iox::dds::FastDDSDataReader::take(
+        uint8_t* const buffer,
```

Comment:
I would also recommend here to use `cxx::not_null` as described above to reduce the error handling.